### DISCUSSION TITLE
USB Direct Supported?

### DISCUSSION THREAD

**Discussion by clee604** [themactep/thingino-firmware#213] [Q&A] [answered]
Hi there, 

I'm currently using a wyze v3 with wz_mini_hacks and evaluating whether I can move over to thingino for my use case. I noticed browsing through the documentation that I can't find any reference to USB Direct or ncm (ie. https://github.com/gtxaspec/wz_mini_hacks/wiki/Configuration-File#usb-direct)

Searching through the code, it does seem like the ncm module is loaded; my main question is whether USB Direct is supported with Thingino? (ie. single cable to supply power and ethernet to the camera from a linux/supported host)

**Comment by gtxaspec** [comment, marked as answer]
hi @clee604, its supported if you compile your own build and enable :
![image](https://github.com/user-attachments/assets/a64cf5fc-56c5-43ea-a8b7-e8fe15ea4a21)

**Reply by clee604** [reply]
ah, so it's not configurable in the webui yet? Requires separate build from the precompiled firmware available?

**Reply by gtxaspec** [reply]
Correct

**Reply by clee604** [reply]
Where does the screenshot you have above come from? I followed instructions on using the docker container script to compile the firmware but there weren't any options aside from selecting the device in the script.

**Reply by gtxaspec** [reply]
`make menuconfig` 
https://github.com/themactep/thingino-firmware/wiki/Development

**Reply by clee604** [reply]
sorry, another question... does this mean I have to compile own build every time I want to update as well or is it just the initial install (ie. does sysupgrade work afterwards?)

**Reply by gtxaspec** [reply]
Sysupgrade will update to the latest compiled image from github, so you would lose your customizations.   To update from your own repo, run git pull, then compile again.  You can update over the air via make uodate_ota IP=x... See the wiki for more info.

**Reply by clee604** [reply]
After messing with the make menuconfig and having repeated failures getting it to compile (this might've been my PC, not sure).. I resorted to adding
```
BR2_PACKAGE_USBNET=y
BR2_PACKAGE_USBNET_USB_DIRECT_NCM=y
```
to local.fragment and running the Guided Compilation script and it compiled successfully. I checked the '.config' file in the output folder and it shows the two above lines set. Any issues doing it this way?

*edit* Didn't seem to work.. the firmware works fine.. but seems to work the same as before (ie. wifi captive portal, connects to wifi and doesn't show as a usb-ncm device to host)

**Reply by clee604** [reply]
I see in the code that 
`# CONFIG_USB_CONFIGFS_NCM is not set`

Is not set by default, do I need to add this paramater to local.fragment with '=y' for USB-NCM to work?

**Reply by gtxaspec** [reply]
you shouldn't need that symbol.
```
BR2_PACKAGE_USBNET=y
BR2_PACKAGE_USBNET_USB_DIRECT_NCM=y
```

will set all the required kernel config symbols.

run `dmesg` ... you should see something like `g_ncm gadget`

**Reply by clee604** [reply]
it looks like 'ncm' doesn't exist in the dmesg and it is loading the usb-eth drivers in the usbcore entries?
```
root@Thingino1 ~# dmesg | grep -iC 3 "ncm"
root@Thingino1 ~# dmesg | grep -iC 3 "usb"
set gpio strength: 37-2
bio: create slab <bio-0> at 0
jz-dma jz-dma: JZ SoC DMA initialized
usbcore: registered new interface driver usbfs
usbcore: registered new interface driver hub
usbcore: registered new device driver usb
 (null): set:249  hold:250 dev=100000000 h=500 l=500
cfg80211: Calling CRDA to update world regulatory domain
Switching to clocksource jz_clocksource
--
dwc2 dwc2: Keep PHY ON
dwc2 dwc2: Using Buffer DMA mode
dwc2 dwc2: Core Release: 3.00a
dwc2 dwc2: DesignWare USB2.0 High-Speed Host Controller
dwc2 dwc2: new USB bus registered, assigned bus number 1
hub 1-0:1.0: USB hub found
hub 1-0:1.0: 1 port detected
dwc2 dwc2: DWC2 Host Initialized
dwc2 dwc2: enter dwc2_gadget_plug_change:2588: plugin = 1 pullup_on = 0 suspend = 0
--
0x000000050000-0x000001000000 : "upgrade"
0x000000000000-0x000001000000 : "all"
SPI NOR MTD LOAD OK
usbcore: registered new interface driver asix
usbcore: registered new interface driver cdc_ether
i2c /dev entries driver
jz-wdt: watchdog initialized
TCP: cubic registered
--
jzmmc_v1.2 jzmmc_v1.2.1: vmmc regulator missing
jzmmc_v1.2 jzmmc_v1.2.1: register success!
exFAT: Version 1.2.9
usbcore: registered new interface driver usbserial
usbcore: registered new interface driver ch341
usbserial: USB Serial support registered for ch341-uart
usbcore: registered new interface driver cp210x
usbserial: USB Serial support registered for cp210x
request spk en gpio 63 ok!
jz_codec_register: probe() successful!
cgu_set_rate, parent = 1392000000, rate = 2048000, n = 10875, reg val = 0x01002a7b
```

I guess this means my compile didn't set the ncm module for some reason? Does the guided compilation tool not use local.fragment options? The config file in the output folder does have my flags set though....

**Reply by gtxaspec** [reply]
check `~/output/<profilename>/.config`

see if it contains
```
BR2_PACKAGE_USBNET=y
BR2_PACKAGE_USBNET_USB_DIRECT_NCM=y```

**Reply by clee604** [reply]
It is set, do I need to disable wifi?
```
BR2_PACKAGE_USBNET=y
# BR2_PACKAGE_USBNET_ASIX is not set
# BR2_PACKAGE_USBNET_ASIX_179_178A is not set
# BR2_PACKAGE_USBNET_CDC_ETHER is not set
BR2_PACKAGE_USBNET_USB_DIRECT_NCM=y
# BR2_PACKAGE_USBNET_R8152 is not set
# BR2_PACKAGE_USBNET_RNDIS is not set
# BR2_PACKAGE_USBNET_WWAN_SERIAL is not set
```

**Reply by gtxaspec** [reply]
hmm...try running the compile again, run `make clean all fast`

**Reply by clee604** [reply]
I'm using the './user-menu.sh' to compile as I'm not familiar enough yet with compiling and command line. I'm going to flush the docker container and start again.

When I tried to run the compile using the 'main menu' (#3 in screenshot) instead of the 'guided compilation' (#2 in screenshot) though, I get this error.
```
Building using unknown with uclibc libs for GCC 5.4.0 toolchain from 1.1.6 SDK
make: *** No rule to make target 'make'.  Stop.
```
![image](https://github.com/user-attachments/assets/ded5523d-173c-4d7c-a866-7879a0aba2f7)

**Reply by gtxaspec** [reply]
that was a bug in the menu system, i just fixed it, pull the latest sources from the repo.

**Reply by clee604** [reply]
PROGRESS! Thanks @gtxaspec. I see usb0 and the ncm gadget in the dmesg now however it doesn't show up as a device on my host. On wz_mini the camera shows up immediately as a enx* network device so I'm not sure if I'm missing anything else?

1. ~~I'm going to try another cable and another host (I'll disconnect a working wz_mini camera using ncm and connect this one)~~
2. I think there's a bug in make fast in the menu system as well, compilation failed when I used that option but it works when I use the regular make option.

*edit* tried with working host/cable that was connected to a wz_mini camera.. no dice.
dmesg on thingino cam log shows
```
root@Thingino1 ~# dmesg | grep -iC 3 "ncm"
using random host ethernet address
usb0: HOST MAC 1a:24:f6:6a:de:7f
usb0: MAC 1e:d3:4a:d6:cd:b0
g_ncm gadget: NCM Gadget
g_ncm gadget: g_ncm ready
gadget pullup defered, current mode: host, plugin: 1
i2c /dev entries driver
jz-wdt: watchdog initialized
```
Do I need to change 'gadget mode' somehow?

**Reply by gtxaspec** [reply]
try `ifconfig usb0` does anything show up?

**Reply by clee604** [reply]
```
root@Thingino1 ~# ifconfig usb0
usb0      Link encap:Ethernet  HWaddr F2:9E:9D:5B:50:5A
          UP BROADCAST MULTICAST  MTU:1500  Metric:1
          RX packets:0 errors:0 dropped:0 overruns:0 frame:0
          TX packets:0 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:1000
          RX bytes:0 (0.0 B)  TX bytes:0 (0.0 B)
```
The one thing I noticed was on wz_mini the cameras show up with 'lsusb' on the connected host system. With the thingino camera nothing shows up for 'lsusb'

ip link on the thingino camera shows the usb0 interface, but it is 'down'. Trying to set it to up didn't work

**Reply by gtxaspec** [reply]
you'll have to manually bring the interface up, or, you can edit `/etc/network/interfaces.d/usb0`:
```
auto usb0
iface usb0 inet dhcp
pre-up usb-role -m device
pre-up ifconfig usb0 hw ether 00:15:de:ad:be:ef
```
and have it be automatic.

**Reply by clee604** [reply]
what's the right command to bring it up? I've tried
'ifconfig usb0 up'
'ip link set usb0 up'
Those do nothing, and when I tried 'ifup eth0' it says 'ifup: interface usb0 already configured'

**Reply by gtxaspec** [reply]
if you see the example above, you can see you need to set the usb role first
`usb-role -m device`
then you can do `ifup usb0` or `ifconfig usb0 up`

need to update the wiki with this info... in thingino you can dynamically switch USB modes, using the `usb-role` utility

**Reply by clee604** [reply]
Thank you! This did it, it shows up in lsusb on my host now. I added it to my network bridge and pulled an IP via dhcp. 

I'm not familiar with what usb-role does so I'll need to google some documentation on this for future reference. I'm guessing wz_mini had this code baked into the code already?

Thank you again for taking the time to help me with this.

**Reply by clee604** [reply]
Did wifi credentials become requiredin a recent release? I wiped the wifi credentials using 
fw_setenv wlanssid
fw_setenv wlanpass

Rebooted.. and this came up on the camera IP
![image](https://github.com/user-attachments/assets/44d48f7a-d086-4636-bd70-37ceb5560a6c)

**Reply by gtxaspec** [reply]
hmm...good question, could be.  i would create an issue regarding this.

**Reply by clee604** [reply]
it's late and i'm not sure if it's because I messed up the configuration elsewhere...

I'll do some more testing tomorrow and create a new issue if I can reproduce it. Thanks again!

**Reply by gtxaspec** [reply]
so we currently check for `eth0`, and if present, the portal is disabled... we should probably extend this to `usb0` as well.  create an issue later if you remember to remind me =D

**Reply by clee604** [reply]
Do you keep wifi enabled on your cameras connected by usb-ncm? Would've assumed you would encounter it once you turn off wifi and connected by usb0 only.

I found that my network bridge broke on my linux host machine though, it seems setting usb0 hw ether only sets one of the mac addresses ('MAC').  The 'HOST MAC' is randomly generated on camera boot, and this means my linux host sees a new network interface name every time (ie. enxdea9d3ef49cb becomes enx{new host mac address on boot} )

```
root@Thingino1 ~# dmesg | grep -iC 3 "ncm"
using random host ethernet address
usb0: HOST MAC de:a9:d3:ef:49:cb
usb0: MAC d2:23:c3:ae:1c:6a
```

Googling how to make this static, I've found a few examples of others with this issue on Pi Zero's. They suggest adding parameters when g_ether module is loaded. I don't see the same configuration files on the thingino camera to do this.

The pi examples:
`modules-load=dwc2,g_ether g_ether.host_addr=00:22:82:ff:ff:20 g_ether.dev_addr=00:22:82:ff:ff:22`
`modprobe g_cdc dev_addr=02:9a:a2:35:63:0e host_addr=d2:d2:c7:d3:ed:90`

On Thingino:
`modprobe: module g_cdc not found in modules.dep`

TDLR: I can't figure out how to keep the host mac of usb0 from changing on every reboot, is there a setting or flag somewhere?

**Reply by gtxaspec** [reply]
In thingino, we build the NCM module into the kernel, so you can't set the host mac.

I dont have an issue with this on OpenWRT.  On your linux machine, there should be a way to assign the interface by USB port, so the interface name doesn't change.  You'll have to do some research on how to set this up, using udev for example, but it should be doable.

**Reply by clee604** [reply]
I guess there's no way to turn off the randomization? I don't necessarily need to set it to a custom value.

On the host side, I saw this post:
https://www.reddit.com/r/Proxmox/comments/xzopay/usb_ethernet_adapter_changing_names_breaking/

It suggests using
`[Match] Driver=ax88179_178a` and then setting the link name manually.

With this method I think it'll work with a single camera attached to the host, but then I'm not sure how it would work if I had multiple cameras since the match is on the driver and not a MAC?

**Reply by clee604** [reply]
Also, is the NCM module different on Thingino compared to wz_mini? The mac isn't randomized on wz_mini using the same host/setup.

**Reply by gtxaspec** [reply]
Same kernel version, same code, in wz_mini we load it as a kernel module, where you can pass arguments.  When compiled into the kernel, there's no way to pass arguments to it (i've tried, the driver ignores them via the sysfs interface) so you have to do the fix on whatever device you are plugging it into, not on the camera itself.  

you need to use udev to set the interface name by usb port, so it won't matter if you have multiple cameras.

**Reply by gtxaspec** [reply]
Some info:

### 1. **Lock the Interface Name Using `udev` Rules**

You can create a persistent network interface name based on the USB port where the device is plugged in. `udev` rules allow you to define custom names for devices based on their physical attributes (like the USB port).

#### Steps:
1. **Identify the USB Device Details**:
   Run the following command to get details about your USB device, including the physical path:

   ```bash
   udevadm info -a -p /sys/class/net/<interface_name> | grep -m 1 "KERNELS=="
   ```

   Replace `<interface_name>` with the current name of your USB-NCM device (e.g., `enxdea9d3ef49cb`). This will output something like:

   ```
   KERNELS=="3-1"
   ```

   This shows the device is connected to USB bus 3, port 1.

2. **Create a `udev` Rule**:
   Create or edit a new `udev` rule file in `/etc/udev/rules.d/`. For example:

   ```bash
   sudo nano /etc/udev/rules.d/99-usb-net.rules
   ```

   Add a rule to lock the interface name based on the USB bus and port:

   ```bash
   SUBSYSTEM=="net", ACTION=="add", KERNELS=="3-1", ATTR{type}=="1", NAME="usbnet0"
   ```

   - `KERNELS=="3-1"`: This is the USB bus and port (from the previous command).
   - `ATTR{type}=="1"`: Ensures it's a network device.
   - `NAME="usbnet0"`: This assigns a persistent interface name `usbnet0` for the device connected to this USB port.

3. **Reload `udev` Rules**:
   Reload `udev` rules and reconnect the USB device:

   ```bash
   sudo udevadm control --reload-rules
   sudo udevadm trigger
   ```

   After this, the network interface should always appear as `usbnet0` (or whatever name you chose) when connected to that specific USB port, regardless of the MAC address.

**Reply by clee604** [reply]
Issue created here for the usb0 portal issue - https://github.com/themactep/thingino-firmware/issues/273

Following the instructions above I was able to successfully set the interface name to usbnet0 using the port location. Using a proxmox machine I'm able to set up the usbnet0 interface and bridge it using a linux bridge. The camera then pulls an IP via dhcp from my router on my network.

I am unable to disable wifi completely though, I've confirmed the usb0 bug above.

**Reply by gtxaspec** [reply]
try
`fw_setenv disable_wlan true` then reboot

**Reply by clee604** [reply]
that works, thanks!

**Reply by clee604** [reply]
this is probably more of a standard linux question but i haven't been able to get 'allow-hotplug' type functionality to work with usbnet0. 

Everything works when the host starts from boot; however if the camera reboots while the host is already running, usbnet0 interface is DOWN until I manually use 'ifup usbnet0'

i have this in my /etc/network/interfaces file
```
allow-hotplug usbnet0
iface usbnet0 inet manual
        up /sbin/ip link set $IFACE master vmbr0
```

I'm guessing I need to modify something in the udex rules?
